### DISCUSSION TITLE
HDF/NetCDF in the cloud

### DISCUSSION THREAD

**Discussion by betolink** [earthaccess-dev/earthaccess#251] [General]
Based on the latest discussions on Openscapes about HDF in the cloud, I think there should be a more in-depth study of the state of things, in a way a "follow up" to this post from Matt Rocklin https://matthewrocklin.com/blog/work/2018/02/06/hdf-in-the-cloud 

`earthaccess` uses `fsspec` for IO but the HDF libraries usually used to read these files(`h5py`, `pyhdf`) were not designed for concurrent access via python file-like objects and IO latency for data in the cloud is really problematic. Here is a list of some of the access patterns used for HDF data in remote storage systems:

* HDF + h5py (via fsspec) -> this is what `earthaccess` uses, it works but is really slow(not necessarily fsspec's fault), see: [h5py sequential reads in the cloud](https://github.com/h5py/h5py/issues/2001) it may also run into [deadlock issues](https://github.com/h5py/h5py/issues/2019) when multiple files are accessed. 
* HDF + [Kerchunk](https://github.com/fsspec/kerchunk) -> seems to work well for gridded NetCDF files, it would be great to have example notebooks and see if it can handle more complex datasets like Icesat's ATL03
* HDF + native S3 driver -> doesn't seem to work for [NASA data ](https://github.com/h5py/h5py/issues/2072) and uses sequential reads too.
* HDF + [H5Coro](https://slideruleearth.io/web/rtd/user_guide/H5Coro.html) -> Only for a subset of HDF files, ~~no stand-alone client library~~ new development!!, H5Coro has a [stand alone client](https://github.com/ICESat2-SlideRule/h5coro) and could be a great drop-in replacement for H5Py on certain cases 
* HDF + OpenDAP = it may work but in practice we delegate the IO and file conversion to OpenDAP and of course we need server side services running. Not all data is supported and/or backed by OpenDAP.

Ideally we could come up with a better understanding of these access patterns: their capabilities, bottlenecks and potential workarounds.

**Comment by weiji14** [comment]
There's been some work done during the ICESat-2 Hackweek 2023 to benchmark reading ICESat-2 ATL03 data from an s3 bucket using different libraries and file formats. See:

- Repository - https://github.com/ICESAT-2HackWeek/h5cloud
- Final presentation - https://docs.google.com/presentation/d/1Hsq1W_X_kbL80tDk8XSozqn1OMcxuv7W8aCZZ2n1ro0/edit#slide=id.g267a8d198f4_1_6

Preliminary results from https://github.com/ICESAT-2HackWeek/h5cloud/commit/1f3441190951e5a2da74611f1196a657db7035bd:

![Comparison of reading an ICESat-2 ATL03 file using h5coro, h5dataframe, h5py, xarray](https://raw.githubusercontent.com/ICESAT-2HackWeek/h5cloud/1f3441190951e5a2da74611f1196a657db7035bd/notebooks/arr_mean_bar_plot.png)

Observations:
- The h5repack format (still HDF5, but optimized chunks) may be faster for newer cloud-optimized libraries like h5coro and kerchunk, but slower for 'classic' libraries like h5py (though it may be due to the rechunking scheme of the h5repack files we tested).
- H5coro and xarray-kerchunk both look like fast and promising areas to continue development. Those libraries are still 'young' though, and generally needs more work to support more HDF5 data structures (e.g. 2D variables for h5coro and nested groups in kerchunk are still unsupported?).

Caveats: Note that the above ATL03 dataset is a **non-gridded point-cloud**-like dataset, and may not apply to other HDF5 data structures, so be careful when generalizing this to **all** HDF5 data files.

---

## Next steps

Some of these are outlined by @abarciauskas-bgse at https://github.com/ICESAT-2HackWeek/h5cloud/issues/18:

- Cloud-optimized data formats:
  - Complete benchmarking of various file formats like kerchunk, h5repack, Zarr, COPC, etc
- Libraries:
  - 2D table-based:
    -  [gedi_subsetter's `H5DataFrame`](https://github.com/MAAP-Project/gedi-subsetter/blob/0.6.0/src/gedi_subset/h5frame.py#L10) (subclassed from `pandas.DataFrame`) that uses `h5py` as the driver, worked on by @chuckwondo et al. Could possibly be extracted as standalone library from GEDI.
  - n-dimensional:
    - Investigate adding new NetCDF4/HDF5 backends to `xarray` besides the current [`h5netcdf`](https://h5netcdf.org):
      - [h5coro](https://github.com/ICESat2-SlideRule/h5coro) (C++ based) - WIP by @jpswinski and @rwegener2 at https://github.com/ICESAT-2HackWeek/xarray (see the h5coro-backend- branches)
      - [hidefix](https://github.com/gauteh/hidefix) (Rust-based) - @gauteh is having an ongoing discussion at https://github.com/pydata/xarray/issues/7446

**Reply by alex-s-gardner** [reply]
@martindurant might be interested in weighing in on this.

**Reply by alex-s-gardner** [reply]
Where all tests run utilizing a comparable number of threads?

**Reply by martindurant** [reply]
Thanks for the ping. Some thoughts.

- it's great to see that kerchunk produces times roughly equal to the very best case even when the target HDF files have not been processed at all. In fact, it is the best for non-repacked, and equal for repacked. Not having to reprocess archival data is the main selling point of kerchunk.
- I have not had much time to look at the benchmark code, but it is hard to produce runs that are fair to all the technologies involved and indeed one would expect some stacks to do better at certain jobs
- kerchunk is not limited in principle to single group netCDF-style data; however xarray *is*. Kerchunk's combining logic assumes CDF, since that most clearly defines which variables are coordinates and how other variables depend on them, but custom reference combinations could be made for any set of HDFs and accessed via zarr (+ dask.array). We would happily see such code in kerchunk.
- similarly, kerchunk has not put much work into tabular data, but there is no reason it cannot be extended. We have thought about feather2, particularly (and maybe reading HDF tables via the arrow API rather than zarr) and even parallel CSV-in-gzip. Of course, the great majority of non-CSV tabular file data is parquet, which already performs pretty well.
- It seems these are running without additional parallelism on top (i.e., dask), which of course can make a big difference. The potential for GIL blocking was the whole reason for developing the experimental rfsspec alongside kerchunk, for instance.


Specifically on the kerchunk benchmark ([xarray_arr_mean.py](https://github.com/ICESAT-2HackWeek/h5cloud/blob/main/h5tests/xarray_arr_mean.py)):
- `[self.open_reference_ds(file) for file in self.files]` - this is, then, a single logical dataset and you could have made a master reference set, reducing the number of calls to read the metadata and duplicated arrays (e.g., coordinates) that xarray is reading each time
- `xr.open_dataset(.., group='gt1l/heights')` similarly, kerchunk allows you to only keep the references you need and rename them if you like, without editing the original files. Here, you are reading all references for all groups, even though you only use a few of them. Of course, this may not make a big difference...

**Reply by abarciauskas-bgse** [reply]
Thanks for these thoughts @martindurant

>It seems these are running without additional parallelism on top (i.e., dask), which of course can make a big difference. The potential for GIL blocking was the whole reason for developing the experimental rfsspec alongside kerchunk, for instance.

It's a good point the scope of these tests was performance without a cluster for parallelizing operations.

I also agree that "hard to produce runs that are fair to all the technologies involved". I am guessing our aim in benchmarking is to learn about the bottlenecks and limitations or advantages of different options, to help inform future data providers and users. In practice, I think we are finding that benchmarking is more of a helpful way to diagnose issues ("oh, why is this method or task slower, that's not what we expected...", for example) than to arrive at a "winner".

With respect to `[self.open_reference_ds(file) for file in self.files]` - I think it is our intention to create a MultiZarr reference (consolidate metadata) but we didn't get the chance in our limited time frame.

I did _not_ know about kerchunk's ability to limit the references needed, so I will look into that for optimizing in the future. However, I imagine we wouldn't want to limit users' ability to access different groups either.

**Reply by abarciauskas-bgse** [reply]
@alex-s-gardner As far as I am aware (although I haven't dug into the underlying library modules used, xarray, h5coro, gedi_subsetter and h5py) no multithreading was used. I do know that the tests were run on a cryocloud JupyterHub instance with an allocated 32-40GB Memory and 4-8 CPU.

**Reply by rwegener2** [reply]
Just wanted to add that h5coro implements multithreading under the hood. That is one of the two major advantages it has over h5py. The other is that it caches the structure of the file as it traverses it, I suppose building a sort of "on the fly" internal kerchunk.

**Reply by gauteh** [reply]
Hidefix also does parallell reading, as well as some parallell decoding (which turns out to be a significant part of xarray load time).

**Reply by martindurant** [reply]
(there is also talk about parallelising decoding in zarr/numcodecs, particularly for GPUs, but this is not yet ready; obviously this is less relevant in dask workflows)

**Reply by betolink** [reply]
Thanks to @jonm3d we now have a tool to rapidly visualize HDF/NetCDF file's structure https://github.com/jonm3D/h5xray. This tool will be very handy when we get to benchmark access to HDF data in cloud storage.

**Comment by betolink** [comment]
## [Cloud optimized HDF5 works](https://gist.github.com/betolink/b545c364f80882c113b8cc27b763c729)!

Well, seems like cloud-optimized HDF5 is indeed possible and the mixed numbers [we initially got](https://github.com/nsidc/earthaccess/discussions/251#discussioncomment-6715494) were the result of not understanding the default IO and caching behavior of `h5py` and `fsspec`. With a better understanding of these mechanisms, we are getting consistently better performance regardless of the file size.

There are some caveats, `h5py` and `fsspec` need to "know" in advance if we are dealing with a cloud optimized HDF5 and if so, tune their behavior to align it with the chunking strategy used to repack the files in question. 
The main caveat of course is that reprocessing entire collections is something that if it happens, will not happen in the short term but these numbers give me hope.

### Exhibit A) 

**Using h5py to calculate a mean  (~47000000  data points) in a 7GB file in-region (us-west-2)**

![Screenshot from 2024-02-05 23-51-53](https://github.com/nsidc/earthaccess/assets/717735/7e6eb7d2-5228-48df-a684-c20d3ae0f7a2)


**Observations**
* h5py is really efficient once it has the file metadata, the performance gain is not dramatic here but if we look the total requests... there is definitely a monetary factor for the data provider.
* Total data transfer could be a bit inaccurate when we use fsspec as some of it comes from the cache, we need a mechanism to separate cache hits vs real data transfers in the logs. Or verify that cache hits are not counted in the reads.

### Exhibit B)

**Doing the same operation using Xarray, in-region access on the same 7GB file**

![Screenshot from 2024-02-05 23-59-34](https://github.com/nsidc/earthaccess/assets/717735/99260436-6e34-4539-b2d6-ec9cee41a770)

**Observations**
* Xarray does by default a lot of CF decoding and that touches a lot of the files. We could turn this off but we don't want to figure out how to interpret each variable (scale, fill etc etc). 
* The overhead of decoding metadata is very noticeable when we open one of the ATL03 granules.
* Using cloud optmized files helps even more in the Xarray access pattern since a lot of the data needed for the decoding needs the metadata structures.

Now let's see what happens when we try to do the same operation over 2 files but this time **out of region**... 

### Exhibit C)

**Out of region access**

![Screenshot from 2024-02-05 18-08-09](https://github.com/nsidc/earthaccess/assets/717735/b1be45e2-e36a-45b9-80d7-0f424210bcf9)

**Observations**
* This scenario is not possible with the S3 distribution for NASA data (restricted to in-region access). If not throttled this would be similar to the performance on HTTPS through NASA's CloudFront.
* This is what cloud-optimized HDF5 could bring in performance and cost savings for the data providers with data in S3, remember each GET request to a bucket costs >0.0 USD. 
* We are using h5py which we know is not multi-threaded, there is room for improvement and if the HDF group gets enough support to keep modernizing the library, we could have a stack in the no distant future where cloud-optimized HDF5 will work the same way COGs work in the raster world. 
* If NISAR and ICESat-2 start producing cloud-optimized datasets, I see the case for `earthaccess.smart_open()` where the library will inspect and align the IO parameters to the strategy in these datasets.


We are working on a set of recommendations specifically for ICESat-2 data and we'll share more findings once we have the document ready.

**Reply by dcherian** [reply]
> Xarray does by default a lot of CF decoding and that touches a lot of the files. We could turn this off but we don't want to figure out how to interpret each variable (scale, fill etc etc).

Xarray *has* to do this on a per-file basis by default because there is no guarantee that the same variable is encoded the same way in different files (`time` is particularly bad, the offsets tend to change). This is the major problem of the "multiple-netcdf-files-constitute-a-dataset" paradigm, and why a true Zarr store is better (there is only one copy of the metadata). Important metadata is replicated in each file with no consistency guarantees. So this complexity leaks in to kerchunk too.

If `earthaccess` knew that the files were all consistent, you can set `decode_cf=False` and call `xr.decode_cf(ds)` after assembling in to a single dataset.

**Reply by martindurant** [reply]
> why a true Zarr store is better (there is only one copy of the metadata). Important metadata is replicated in each file with no consistency guarantees. So this complexity leaks in to kerchunk too.

Well, kerchunk hopes to decode once, so that reading the created reference set is one-step and doesn't access the constituent data files. Yes, we've had to spend quite some effort getting concatenation in the presence of different CF time "units" fields right.

**Reply by dcherian** [reply]
Yes certainly kerchunk is a major performance enhancement, but I'm hoping to convince our data center friends here that even if you can, perhaps you shouldn't when it comes to HDF in the cloud.

**Reply by martindurant** [reply]
Of course, if straight-to-zarr is an option, that's different. Or using normal time units. Or at the very least having a consistent epoch... Data centres do still seem to like using the most established format and conventions, and also the ability to just download some part of the data as self-contained, self-describing files.

**Reply by betolink** [reply]
Hopefully this was not interpreted as a critique to Xarray decoding, I think the overall objective of this topic is to understand different access patterns and their tradeoffs when it comes to accessing archival data in the cloud. 

I see cloud-optimized HDF5 as a multiplier for approaches like kerchunk or even pangeo-forge pipelines (if they have to transform HDF to Zarr). If used, generating the references for kerchunk can benefit from already consolidated metadata in the HDF5 file, anecdotally I noticed a ~5x improvement (PR coming to kerchunk). I also wonder if we should copy/follow up this valuable conversation on this thread in [Pangeo](https://discourse.pangeo.io/t/whats-next-data-management/3891/5)

**Reply by martindurant** [reply]
> PR coming to kerchunk

Looking forward to it :)

**Comment by zequihg50** [comment]
Hi,

Great results! Have you attempted to use xarray and dask with the processes scheduler? I'm experimenting with this, but for some unknown reason, the volume of data (Total Req) is huge, leading to longer processing times. For instance:

```python
import xarray
import dask
import fsspec

dask.config.set(scheduler="processes")

URL = "..."

fs = fsspec.open(URL, cache_type="blockcache", block_size=8*1024*1024)

ds = xarray.open_dataset(
    fs.open(),
    engine="h5netcdf",
    driver_kwds={
        "page_buf_size": 32*1024*1024,
        "rdcc_nbytes": 8*1024*1024
    }).chunk({...}) # align with file chunking

ds["variable"].mean().compute(num_workers=8)
```

**Reply by martindurant** [reply]
The best caching scheme for HDF5 is usually "first", especially if you set the block size big enough to contain all front matter of the file. Of course, the processes cannot share this in-memory cache, so it may well not be doing too much that is helpful for you. You might also want to try "blockcache::" prepended to your URL for (shared) on-disk block caching, but you would need to create your store directory first (`cache_storage=` kwarg to the caching filesystem).

**Reply by betolink** [reply]
> the process cannot share the cache

Maybe this is what's happening, I'd try using Martin's suggestion for a shared cache on disk. Another thing, the `blockcache` works better if the files were cloud optimized first and the page size aligns with that of the cache, I'm not sure if the files you're opening have that. If they are public maybe we can test some of these approaches.

**Reply by zequihg50** [reply]
Thank you both for the insightful comments. Indeed, utilizing the "first" caching scheme improves things, but the volume of data read remains huge when employing Dask processes. I will experiment with the disk-shared cache and provide feedback if I make any progress. While HDF5 paged aggregation notably enhances performance, it appears from the numbers you've presented that there is a substantial increase in Total Req Bytes for all Exhibits, which can be exacerbated in some cases, such as process parallelism. Anyway, a lot of interesting stuff happening :)

**Reply by betolink** [reply]
@zequihg50 I forgot to mention, the total requested bytes in the graph is a bit misleading because fsspec logs do not track cache hits (PR coming!) meaning the actual requested bytes is way less than the ones in the charts I posted.
